lib/spack/spack/test/stage.py

```diff
@@ -902,8 +917,91 @@ def test_diystage_preserve_file(self, tmpdir):
         with open(readmefn) as _file:
             _file.read() == _readme_contents
 
+    def test_create_cmakebuildstage(self, cmake_ext_build, mock_stage_archive):
+        """Test that cmake external build dir is created during stage creation"""
+        archive = mock_stage_archive()
+        with CMakeBuildStage("jk6dj", archive.url, name=self.stage_name) as stage:
+            # external build dir should already have been created at this point
+            # ensure symlink to build dir exists
+            assert str(stage.root_stage_context.name) == "spack-build-jk6dj"
+            assert "spack-build-jk6dj" in os.listdir(stage.path)
+            # check symlink points to the right place
+            assert str(stage.root_stage_context.resolve()) == str(cmake_ext_build / "a")
+            # ensure ext build dir was set honoring config
+            assert spack.config.get("config:cmake_ext_build") == str(stage._remote_stage.parent)
+
+            # ensure actual external build dir is present
+            assert stage._remote_stage
+            assert str(stage._remote_stage.name) in os.listdir(cmake_ext_build)
+            assert str(stage._remote_stage) == str(cmake_ext_build / "a")
+
+        # ensure ext build stage is destroyed on cleanup
+        assert not "a" is os.listdir(cmake_ext_build)
+
+    def test_create_cmakebuildstage_non_empty(self, populated_cmake_ext_build, mock_stage_archive):
+        """Test that cmake external build dir can handle ext build dir with other builds
+        already present"""
+        archive = mock_stage_archive()
+        with CMakeBuildStage("jk6dj", archive.url, name=self.stage_name) as stage:
+            # external build dir should already have been created at this point
+            # ensure symlink to build dir exists
+            assert "spack-build-jk6dj" in os.listdir(stage.path)
+            # ensure actual external build dir is present
+            assert str(stage._remote_stage.name) in os.listdir(populated_cmake_ext_build)
+            # ensure remote stage has correct name based on already present stages
+            # the actual name matters less here than avoiding colisions between already
+            # present build dirs
+            assert str(stage._remote_stage) == str(populated_cmake_ext_build / "b")
+
+        # ensure ext build stage is destroyed on cleanup
+        assert not "b" is os.listdir(populated_cmake_ext_build)
```

Comment:
I recommend grabbing `build_directory` from the `Stage` object vs. inferring the name based on how you know the underling generator works (if someone meddles with that, this check will silently become useless)

Reply:
Because at this point the `build_directory` attribute is `None`. When the stage exists the context manager, its ext build stage associated properties are deallocated as they don't point to anything anymore. Should we refactor that behavior?